dali/pipeline/operator/builtin/input_operator.h

```diff
@@ -439,18 +400,30 @@ class InputOperator : public Operator<Backend>, virtual public BatchSizeProvider
     auto tl_elm = GetEmptyOutputBatch(std::move(data_id));
     bool copied_shared_data = false;
 
-    if (batch.IsContiguous()) {
+    // We can share only contiguous tensor lists that are stored on the same device.
+    if (batch.IsContiguousInMemory() && batch.device_id() == device_id_) {
       auto &batch_owner = unsafe_sample_owner(const_cast<TensorList<SrcBackend> &>(batch), 0);
-      tl_elm.front()->ShareData(batch_owner, batch.nbytes(), batch.is_pinned(), batch.shape(),
+      tl_elm->data.ShareData(batch_owner, batch.nbytes(), batch.is_pinned(), batch.shape(),
                                 batch.type(), batch.device_id(), batch.order());
       zero_copy_noncontiguous_gpu_input_ = true;
     } else {
-      // it is not contiguous so we need to copy
-      tl_elm.front()->Copy(batch, order, use_copy_kernel);
-      std::list<uptr_cuda_event_type> copy_to_storage_event;
-      copy_to_storage_event = copy_to_storage_events_.GetEmpty();
-      CUDA_CALL(cudaEventRecord(*copy_to_storage_event.front(), order.stream()));
-      copy_to_storage_events_.PushBack(copy_to_storage_event);
+      // Do not overwrite the buffer it if shares data.
+      if (tl_elm->data.shares_data())
+        tl_elm->data.Reset();
+      tl_elm->data.Copy(batch, order, use_copy_kernel);
+      int device_id = order.is_device() ? order.device_id() : tl_elm->data.device_id();
+      cudaEvent_t event = tl_elm->GetCompletionEvent(order.device_id());
+
+      if (order.device_id() != device_id_ &&
```

Comment:
This condition is unreadable, can you make it a function with readable name? Like is is_ambiguous_order in access_order.cc?

Reply:
Since it's done once per iteration, I think we can afford to just always go with the `DeviceGuard`. I couldn't make such assumptions in `access_order.cc`, but here it's quite different.

Reply:
I always add DeviceGuard. It takes 80ns on my machine and that's done once per iteration, so it's rather cheap.